Using msbuild over xbuild causes issues ( at least on Mono 4.8.x)

cc @baronfel 

Recent FAKE (> 4.53.0?)  has caused a regression on Linux/macOS with Mono 4.8 and Mono latest (currently 4.8.1).  It invokes MSBuild instead of XBuild on these platforms ([this PR](https://github.com/fsharp/FAKE/pull/1488#issuecomment-298158602)).

That's fine, but I believe MSBuild on Unix/Mono has problems, at least before Mono 5.0.0 (yet to be released).  For example, when compiling FSharp.Compiler.Service with (latest FAKE and FSharp.,Compiler.Tools](https://travis-ci.org/fsharp/FSharp.Compiler.Service/jobs/227043065) using this:

```
  /Library/Frameworks/Mono.framework/Versions/Current/Commands/msbuild  /Users/travis/build/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.sln /t:Build    /p:RestorePackages="False" /p:Configuration="Release" /p:TargetFrameworkVersion="v4.5" /logger:Fake.MsBuildLogger+ErrorLogger,"/Users/travis/build/fsharp/FSharp.Compiler.Service/packages/FAKE/tools/FakeLib.dll"
```

I get

```
/Users/travis/build/fsharp/FSharp.Compiler.Service/packages/FSharp.Compiler.Tools/tools/Microsoft.FSharp.Targets(125,13): 
error MSB4131: The "ResourceFilesWithManifestResourceNames" parameter is not supported by 
the "CreateFSharpManifestResourceName" task. Verify the parameter exists on the task, and it 
is a gettable public instance property. 
[/Users/travis/build/fsharp/FSharp.Compiler.Service/src/fsharp/FSharp.Compiler.Service/FSharp.Compiler.Service.fsproj]
```

This seems to be an MSBuild on Unix and Mono issue.

I suspect we should be avoiding defaulting to MSBuild on Unix, at least on Mono 4.8.x and before.  I've yet to validate if this issue goes away on Mono 5.0.0.

### Repro steps

Consider [this PR](https://github.com/fsharp/FSharp.Compiler.Service/pull/765).  

* [this commit fails to build](https://github.com/fsharp/FSharp.Compiler.Service/pull/765/commits/4d5fa48fba6eda790fdbe65914c22ac9af4571ce).  It uses latest FAKE and invokes ``msbuild`` on Unix

* I'm in the process of validating that [this commit succeeds](https://github.com/fsharp/FSharp.Compiler.Service/pull/765/commits/85d3f960ba5c177a0706f2b3fc55cb59b5d52914).  It uses FAKE 4.53.0 instead of FAKE 4.60.0 and invokes ``xbuild`` on Unix

### Known workarounds

Use FAKE 4.53.0 or before (I'm in the process of validating this)
Perhaps use Mono 5.0.0 (?)

### Related information 

* Unix, Mono 4.8.*

--- a/src/app/FakeLib/MSBuildHelper.fs
+++ b/src/app/FakeLib/MSBuildHelper.fs
@@ -45,19 +45,28 @@ let toDict items =
 let getAllKnownPaths =
     (knownMsBuildEntries |> List.collect (fun m -> m.Paths) |> List.rev) @ oldMsBuildLocations
 
+/// Versions of Mono prior to this one have faulty implementations of MSBuild
+let monoVersionToUseMSBuildOn = System.Version("5.0.0.0")
 
 /// Tries to detect the right version of MSBuild.
 ///   - On all OS's, we check a `MSBuild` environment variable which is either
 ///     * a direct path to a file to use, or
-///     * a directory that contains a file called `msbuild` on non-Windows systems, or `MSBuild.exe` on Windows systems
+///     * a directory that contains a file called 
+///         * `msbuild` on non-Windows systems with mono >= 5.0.0.0, or
+///         * `xbuild` on non-Windows systems with mono < 5.0.0.0,
+///         * `MSBuild.exe` on Windows systems, or
+///     * a tool that exists on the current PATH
 ///   - In addition, on non-Windows systems we check the current PATH for the following binaries, in this order:
-///     * `msbuild`
-///     * `xbuild`
+///     * Mono >= 5.0.0.0: `msbuild`, `xbuild`
+///     * Mono < 5.0.0.0: `xbuild`, `msbuild`
+///     * This is due to several known issues in the Mono < 5.0 implementation of MSBuild.
 ///   - In addition, on Windows systems we
 ///     * try to read the MSBuild tool location from the AppSettings file using a parameter named `MSBuild`, and finally
 ///     * if a `VisualStudioVersion` environment variable is specified, we try to use the specific MSBuild version, matching that Visual Studio version.
 let msBuildExe =
-    /// the value we're given can be a full path to a file or just a directory.
+    /// the value we're given can be a:
+    ///     * full path to a file or
+    ///     * just a directory
     /// if just a directory we can make it the path to a file by Path-Combining the tool name to the directory.
     let exactPathOrBinaryOnPath tool input =
         if FileSystemHelper.isDirectory input && Directory.Exists input
@@ -65,19 +74,27 @@ let msBuildExe =
         else input
 
     let which tool = ProcessHelper.tryFindFileOnPath tool
-    let environVarDir = EnvironmentHelper.environVarOrNone "MSBuild"
+    let msbuildEnvironVar = EnvironmentHelper.environVarOrNone "MSBuild"
 
-    if isUnix
-    then
-        let unixSources = [
-            environVarDir |> Option.map (exactPathOrBinaryOnPath "msbuild")
+    match isUnix, EnvironmentHelper.monoVersion with
+    | true, Some(_, Some(version)) when version > monoVersionToUseMSBuildOn -> 
+        let sources = [
+            msbuildEnvironVar |> Option.map (exactPathOrBinaryOnPath "msbuild")
+            msbuildEnvironVar |> Option.bind which
             which "msbuild"
             which "xbuild"
         ]
-        defaultArg (unixSources |> List.choose id |> List.tryHead) "xbuild"
-    else
+        defaultArg (sources |> List.choose id |> List.tryHead) "xbuild"
+    | true, _ -> 
+        let sources = [
+            msbuildEnvironVar |> Option.map (exactPathOrBinaryOnPath "xbuild")
+            msbuildEnvironVar |> Option.bind which
+            which "xbuild"
+            which "msbuild"
+        ]
+        defaultArg (sources |> List.choose id |> List.tryHead) "xbuild"
+    | false, _ -> 
 
-        let envVarPath = environVarDir |> Option.map (exactPathOrBinaryOnPath "MSBuild.exe")
         let configIgnoreMSBuild =
             if "true".Equals(ConfigurationManager.AppSettings.["IgnoreMSBuild"], StringComparison.OrdinalIgnoreCase)
             then Some ""
@@ -90,12 +107,13 @@ let msBuildExe =
 
             ProcessHelper.tryFindFileInDirsThenPath vsVersionPaths "MSBuild.exe"
 
-        let windowsSources = [
-            envVarPath
+        let sources = [
+            msbuildEnvironVar |> Option.map (exactPathOrBinaryOnPath "MSBuild.exe")
+            msbuildEnvironVar |> Option.bind which
             configIgnoreMSBuild
             findOnVSPathsThenSystemPath
         ]
-        defaultArg (windowsSources |> List.choose id |> List.tryHead) "MSBuild.exe"
+        defaultArg (sources |> List.choose id |> List.tryHead) "MSBuild.exe"
 
 
 /// [omit]
@@ -446,7 +464,10 @@ let MSBuildWithProjectProperties outputPath (targets : string) (properties : (st
             |> Set.unionMany
 
     let setBuildParam project projectParams =
-        { projectParams with Targets = targets |> split ';' |> List.filter ((<>) ""); Properties = projectParams.Properties @ properties project }
+        { projectParams with 
+            Targets = targets |> split ';' |> List.filter ((<>) "")
+            Properties = projectParams.Properties @ properties project
+            MaxCpuCount = Some ( Some (EnvironmentHelper.getMachineEnvironment().ProcessorCount)) }
 
     projects
       |> List.filter (fun project -> not <| Set.contains project dependencies)

--- a/src/app/FakeLib/MSBuildHelper.fs
+++ b/src/app/FakeLib/MSBuildHelper.fs
@@ -466,8 +466,7 @@ let MSBuildWithProjectProperties outputPath (targets : string) (properties : (st
     let setBuildParam project projectParams =
         { projectParams with 
             Targets = targets |> split ';' |> List.filter ((<>) "")
-            Properties = projectParams.Properties @ properties project
-            MaxCpuCount = Some ( Some (EnvironmentHelper.getMachineEnvironment().ProcessorCount)) }
+            Properties = projectParams.Properties @ properties project }
 
     projects
       |> List.filter (fun project -> not <| Set.contains project dependencies)

--- a/src/app/FakeLib/MSBuildHelper.fs
+++ b/src/app/FakeLib/MSBuildHelper.fs
@@ -77,14 +77,14 @@ let msBuildExe =
     let msbuildEnvironVar = EnvironmentHelper.environVarOrNone "MSBuild"
 
     match isUnix, EnvironmentHelper.monoVersion with
-    | true, Some(_, Some(version)) when version > monoVersionToUseMSBuildOn -> 
+    | true, Some(_, Some(version)) when version >= monoVersionToUseMSBuildOn -> 
         let sources = [
             msbuildEnvironVar |> Option.map (exactPathOrBinaryOnPath "msbuild")
             msbuildEnvironVar |> Option.bind which
             which "msbuild"
             which "xbuild"
         ]
-        defaultArg (sources |> List.choose id |> List.tryHead) "xbuild"
+        defaultArg (sources |> List.choose id |> List.tryHead) "msbuild"
     | true, _ -> 
         let sources = [
             msbuildEnvironVar |> Option.map (exactPathOrBinaryOnPath "xbuild")

